includes/payment-methods/class-upe-payment-gateway.php

```diff
@@ -420,6 +420,22 @@ public function process_payment( $order_id ) {
 					throw new Exception( WC_Payments_Utils::get_filtered_error_message( $e ) );
 				}
 
+				$intent_id              = $updated_payment_intent->get_id();
+				$intent_status          = $updated_payment_intent->get_status();
+				$payment_method         = $updated_payment_intent->get_payment_method_id();
+				$payment_method_details = $updated_payment_intent->get_payment_method_details();
+				$payment_method_type    = $payment_method_details ? $payment_method_details['type'] : null;
+				$charge_id              = $updated_payment_intent->get_charge_id();
+
+				/**
+				 * Attach the intent and exchange info to the order before doing the redirect, just in case the redirect
+				 * either does not complete properly, or the Stripe webhook which processes a successful order hits before
+				 * the redirect completes.
+				 */
+				$this->attach_intent_info_to_order( $order, $intent_id, $intent_status, $payment_method, $customer_id, $charge_id, $currency );
+				$this->attach_exchange_info_to_order( $order, $updated_payment_intent->get_charge_id() );
+				$this->set_payment_method_title_for_order( $order, $payment_method_type, $payment_method_details );
```

Comment:
> It looks like the webhook looks up the order using the intent ID, so we'll need to ensure we've added it here.

The webhook has a fallback to get the order by `order_id` if it cannot find it by intent ID, so this doesn't have an affect on the webhook logic 🙂 the main idea behind attaching it at this point was to ensure it's always present on the order even if for some reason the redirect goes wrong (since this metadata is vital to tie the order to the transaction and to process a refund on the order). It may potentially be overkill, since we also attach it in the redirect.

> I wonder: do we also need to also ensure the setup intent ID is present on the order when $payment_needed is false? I didn't see notice any webhooks flying around when I completed a transaction using a setup intent (sub with a free-trial), so perhaps we don't need to worry about this case.

Good question! I will double check as well to see if any webhooks are fired with an order that doesn't require payment - if there's no webhook which processes the order then we should be able to rely on the current logic.